### ISSUE TITLE
Number of broadcasting dimensions is limited

### ISSUE THREAD

**Issue by sjperkins** [tensorflow/tensorflow#1519]
### Environment info

Operating System: Ubuntu 14.04.4

If installed from sources, provide the commit hash: f92e338b178f75b0e7fc79aedc00680a49a982ac
### Steps to reproduce

Running this script:

``` python
import tensorflow as tf

shape_1 = (1, 20, 30, 1, 50)
shape_2 = (30, 1, 30, 20, 50)

A = tf.Variable(tf.random_normal(shape_1))
B = tf.Variable(tf.random_normal(shape_2))

with tf.Session() as S:
    S.run(tf.initialize_all_variables())
    C = S.run(A+B)
```

produces

``` python
tensorflow.python.framework.errors.UnimplementedError: Broadcast between [1,20,30,1,50] and [30,1,30,20,50] is not supported yet.
```
### What did you try?

Reducing the tensor ranks results in success.

``` python
shape_1 = (1, 20, 30)
shape_2 = (30, 1, 30)
```
### Logs or other output that would be helpful

``` python
Traceback (most recent call last):
  File "tf_broadcast_issue.py", line 11, in <module>
    C = S.run(A+B)
  File "/home/sperkins/work/ska/code/tensorflow/_python_build/tensorflow/python/client/session.py", line 332, in run
    result = self._run(None, fetches, feed_dict, options_ptr, run_outputs_ptr)
  File "/home/sperkins/work/ska/code/tensorflow/_python_build/tensorflow/python/client/session.py", line 536, in _run
    feed_dict_string, options, run_outputs)
  File "/home/sperkins/work/ska/code/tensorflow/_python_build/tensorflow/python/client/session.py", line 598, in _do_run
    target_list, options, run_outputs)
  File "/home/sperkins/work/ska/code/tensorflow/_python_build/tensorflow/python/client/session.py", line 620, in _do_call
    e.code)
tensorflow.python.framework.errors.UnimplementedError: Broadcast between [1,20,30,1,50] and [30,1,30,20,50] is not supported yet.
     [[Node: add = Add[T=DT_FLOAT, _device="/job:localhost/replica:0/task:0/cpu:0"](Variable/read, Variable_1/read)]]
Caused by op u'add', defined at:
  File "tf_broadcast_issue.py", line 11, in <module>
    C = S.run(A+B)
  File "/home/sperkins/work/ska/code/tensorflow/_python_build/tensorflow/python/ops/variables.py", line 544, in <lambda>
    setattr(Variable, operator, lambda a, b: Variable._RunOp(operator, a, b))
  File "/home/sperkins/work/ska/code/tensorflow/_python_build/tensorflow/python/ops/variables.py", line 559, in _RunOp
    return getattr(ops.Tensor, operator)(a._AsTensor(), b)
  File "/home/sperkins/work/ska/code/tensorflow/_python_build/tensorflow/python/ops/math_ops.py", line 502, in binary_op_wrapper
    return func(x, y, name=name)
  File "/home/sperkins/work/ska/code/tensorflow/_python_build/tensorflow/python/ops/gen_math_ops.py", line 44, in add
    return _op_def_lib.apply_op("Add", x=x, y=y, name=name)
  File "/home/sperkins/work/ska/code/tensorflow/_python_build/tensorflow/python/ops/op_def_library.py", line 655, in apply_op
    op_def=op_def)
  File "/home/sperkins/work/ska/code/tensorflow/_python_build/tensorflow/python/framework/ops.py", line 2102, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "/home/sperkins/work/ska/code/tensorflow/_python_build/tensorflow/python/framework/ops.py", line 1129, in __init__
    self._traceback = _extract_stack()
```

**Comment by sjperkins** [comment]
#206 might resolve this, but I've logged it separately in case.

**Comment by sjperkins** [comment]
Looks like member `BinaryOp.Compute` in [cwise_ops_common.h](https://github.com/tensorflow/tensorflow/blob/f92e338b178f75b0e7fc79aedc00680a49a982ac/tensorflow/core/kernels/cwise_ops_common.h#L85-L139) only supports broadcasting between 2D and 3D shapes.

**Comment by mrry** [comment]
I think we've limited the number of dimensions to keep binary size down, rather than because it's any more complicated to broadcast on more dimensions. @benoitsteiner, is there some way we could handle higher dimensions in Eigen without another specialization, even if it leads to a slower codepath?

**Comment by benoitsteiner** [comment]
@mrry If I'm not mistaken, the broadcasts we need in the examples above are [30, 1, 1, 20, 1] for the first tensor, and [1, 20, 1, 1, 1] for the second one. For the first tensor, we could collapse the 2nd and 3rd dimensions together, and apply a  [30, 1, 20, 1] broadcast. For the second one, we could similarly reshape the tensor into a 3d tensor by collapsing the last 3 dimensions, and apply a [1, 20, 1] broadcast.

**Comment by buyijie** [comment]
I want to caculate the Kronecker product of two matrix, and 
tf.reshape(matrix_A, [2,1,2,1]) * tf.reshape(matrix_B, [1,3,1,3]) raised the above UnimpementedError.
I circumvented this broadcasting by using tf.tile to explictly broadcast this two matrix. 

Is there an easy way to do this broadcasting in tensorflow, like in Numpy?

**Comment by michaelisard** [comment]
@benoitsteiner any fix should be sure to cover the cases in #3534 as well.

**Comment by tatatodd** [comment]
@benoitsteiner Friendly ping, please update this bug with the current status, thanks!

**Comment by drpngx** [comment]
Fixed by #4378.
